File: tools/crabgen/tests/golden/cpp/full/impl.cpp
// impl.cpp — the application half of this guest: define the impl:: functions
// declared in gen/bindings.hpp. crabgen scaffolds this file ONCE and never
// overwrites it; `crabgen regen` prints any missing function signatures
// instead of editing it (a missing definition is also a LINK ERROR naming
// the symbol when build.sh links the module).
#include "gen/bindings.hpp"

namespace impl {

// echo_everything handles crab:full/kitchen@0.1.0#echo-everything.
// A non-empty .err is a function-level failure (status-1 reply).
crab::Res<gen::Everything> echo_everything(gen::Everything e) {
  return crab::Res<gen::Everything>::fail("unimplemented: echo-everything");
}

// pick_color handles crab:full/kitchen@0.1.0#pick-color.
// A non-empty .err is a function-level failure (status-1 reply).
crab::Res<gen::Color> pick_color(gen::Color c) {
  return crab::Res<gen::Color>::fail("unimplemented: pick-color");
}

// set_perms handles crab:full/kitchen@0.1.0#set-perms.
// A non-empty .err is a function-level failure (status-1 reply).
crab::Res<gen::Perms> set_perms(gen::Perms p) {
  return crab::Res<gen::Perms>::fail("unimplemented: set-perms");
}

// classify handles crab:full/kitchen@0.1.0#classify.
// A non-empty .err is a function-level failure (status-1 reply).
crab::Res<std::string> classify(gen::Shape s) {
  return crab::Res<std::string>::fail("unimplemented: classify");
}

// try_divide handles crab:full/kitchen@0.1.0#try-divide.
// A non-empty .err encodes as the WIT result err case (a normal status-0 reply).
crab::Res<double> try_divide(double num, double den) {
  return crab::Res<double>::fail("unimplemented: try-divide");
}

// maybe_list handles crab:full/kitchen@0.1.0#maybe-list.
// A non-empty .err is a function-level failure (status-1 reply).
crab::Res<std::vector<std::optional<bool>>> maybe_list(std::optional<std::vector<uint16_t>> xs) {
  return crab::Res<std::vector<std::optional<bool>>>::fail("unimplemented: maybe-list");
}

// no_result handles crab:full/kitchen@0.1.0#no-result.
// A non-empty .err is a function-level failure (status-1 reply).
crab::Res<std::monostate> no_result(uint32_t x) {
  return crab::Res<std::monostate>::fail("unimplemented: no-result");
}

// retry handles crab:full/kitchen@0.1.0#retry.
// A non-empty .err is a function-level failure (status-1 reply).
crab::Res<gen::Result<uint32_t, gen::Color>> retry(std::optional<gen::Result<uint32_t, gen::Color>> prev) {
  return crab::Res<gen::Result<uint32_t, gen::Color>>::fail("unimplemented: retry");
}

}  // namespace impl

File: tools/crabgen/tests/golden/go/full/impl.go
// impl.go — the application half of this guest: implement gen.Impl here.
// crabgen scaffolds this file ONCE and never overwrites it; `crabgen regen`
// prints any missing method signatures instead of editing it.
package main

import (
	"fmt"

	"crabcraft.local/full/gen"
)

// App implements gen.Impl: one method per function exported by
// crab:full/kitchen@0.1.0.
type App struct{}

// EchoEverything handles crab:full/kitchen@0.1.0#echo-everything.
// A non-nil error is a function-level failure (status-1 reply).
func (App) EchoEverything(e gen.Everything) (gen.Everything, error) {
	return gen.Everything{}, fmt.Errorf("unimplemented: echo-everything")
}

// PickColor handles crab:full/kitchen@0.1.0#pick-color.
// A non-nil error is a function-level failure (status-1 reply).
func (App) PickColor(c_ gen.Color) (gen.Color, error) {
	return 0, fmt.Errorf("unimplemented: pick-color")
}

// SetPerms handles crab:full/kitchen@0.1.0#set-perms.
// A non-nil error is a function-level failure (status-1 reply).
func (App) SetPerms(p gen.Perms) (gen.Perms, error) {
	return 0, fmt.Errorf("unimplemented: set-perms")
}

// Classify handles crab:full/kitchen@0.1.0#classify.
// A non-nil error is a function-level failure (status-1 reply).
func (App) Classify(s gen.Shape) (string, error) {
	return "", fmt.Errorf("unimplemented: classify")
}

// TryDivide handles crab:full/kitchen@0.1.0#try-divide.
// A non-nil error encodes as the WIT result err case (a normal status-0 reply).
func (App) TryDivide(num float64, den float64) (float64, error) {
	return 0, fmt.Errorf("unimplemented: try-divide")
}

// MaybeList handles crab:full/kitchen@0.1.0#maybe-list.
// A non-nil error is a function-level failure (status-1 reply).
func (App) MaybeList(xs *[]uint16) ([]*bool, error) {
	return nil, fmt.Errorf("unimplemented: maybe-list")
}

// NoResult handles crab:full/kitchen@0.1.0#no-result.
// A non-nil error is a function-level failure (status-1 reply).
func (App) NoResult(x uint32) error {
	return fmt.Errorf("unimplemented: no-result")
}

func init() { gen.SetImpl(App{}) }

// main never runs: the module builds as a wasip1 reactor (the host calls
// _initialize once, then crab_invoke per request).
func main() {}
